File: contract/projects/algosplit/smart_contracts/algosplit/contract.py
from algopy import (
    ARC4Contract,
    Account,
    Bytes,
    Global,
    Txn,
    UInt64,
    gtxn,
    op,
    subroutine,
    urange,
)
from algopy.arc4 import abimethod, Bool, String, UInt64 as ARC4UInt64, DynamicArray, Struct


class Bill(Struct):
    """Bill box structure"""
    creator: Account
    total_amount: UInt64
    member_count: UInt64
    settled_count: UInt64
    is_settled: Bool


class Member(Struct):
    """Member box structure"""
    share: UInt64
    paid: Bool


class Algosplit(ARC4Contract):
    """AlgoSplit - On-chain Split Bill Settlement Smart Contract"""

    def __init__(self) -> None:
        """Initialize contract with bill counter"""
        self.bill_counter = UInt64(0)

    @abimethod()
    def create_bill(
        self,
        members: DynamicArray[Account],
        shares: DynamicArray[ARC4UInt64],
        mbr_payment: gtxn.PaymentTransaction,
    ) -> ARC4UInt64:
        """
        Create a new bill with members and their shares
        
        Args:
            members: List of member addresses
            shares: List of share amounts for each member
            mbr_payment: Payment transaction for box storage MBR
            
        Returns:
            Bill ID
        """
        # Validate inputs
        assert members.length == shares.length, "Members and shares length mismatch"
        assert members.length > 0, "At least one member required"
        
        # Verify MBR payment
        assert mbr_payment.receiver == Global.current_application_address, "MBR payment must go to app"
        
        # Increment bill counter
        self.bill_counter += 1
        bill_id = self.bill_counter
        
        # Calculate total amount
        total_amount = UInt64(0)
        for i in urange(members.length):
            share_amount = shares[i].native
            assert share_amount > 0, "Share must be greater than 0"
            total_amount += share_amount
        
        # Create bill box
        bill_key = self._get_bill_key(bill_id)
        bill = Bill(
            creator=Txn.sender,
            total_amount=total_amount,
            member_count=members.length,
            settled_count=UInt64(0),
            is_settled=Bool(False),
        )
        op.Box.put(bill_key, bill.bytes)
        
        # Create member boxes
        for i in urange(members.length):
            member_address = members[i]
            share_amount = shares[i].native
            
            member_key = self._get_member_key(bill_id, member_address)
            member = Member(
                share=share_amount,
                paid=Bool(False),
            )
            op.Box.put(member_key, member.bytes)
            
            # Add bill to user's participation list
            self._add_bill_to_user(member_address, bill_id)
        
        return ARC4UInt64(bill_id)

    @abimethod()
    def pay_bill(
        self,
        bill_id: ARC4UInt64,
        payment: gtxn.PaymentTransaction,
    ) -> Bool:
        """
        Pay a bill share (grouped with payment transaction)
        
        Args:
            bill_id: The bill ID to pay
            payment: Payment transaction to bill creator
            
        Returns:
            Success status
        """
        bill_id_native = bill_id.native
        
        # Load bill
        bill_key = self._get_bill_key(bill_id_native)
        bill_bytes, exists = op.Box.get(bill_key)
        assert exists, "Bill does not exist"
        bill = Bill.from_bytes(bill_bytes)
        
        assert not bill.is_settled.native, "Bill already settled"
        
        # Load member
        member_key = self._get_member_key(bill_id_native, Txn.sender)
        member_bytes, member_exists = op.Box.get(member_key)
        assert member_exists, "Not a member of this bill"
        member = Member.from_bytes(member_bytes)
        
        assert not member.paid.native, "Already paid"
        
        # Verify payment transaction
        assert payment.sender == Txn.sender, "Payment sender mismatch"
        assert payment.receiver == bill.creator, "Payment must go to bill creator"
        assert payment.amount == member.share, "Payment amount must match share"
        
        # Update member status
        member.paid = Bool(True)
        op.Box.put(member_key, member.bytes)
        
        # Update bill status
        bill.settled_count += 1
        if bill.settled_count == bill.member_count:
            bill.is_settled = Bool(True)
        op.Box.put(bill_key, bill.bytes)
        
        return Bool(True)

    @abimethod(readonly=True)
    def get_bill(self, bill_id: ARC4UInt64) -> Bill:
        """
        Get bill details
        
        Args:
            bill_id: The bill ID
            
        Returns:
            Bill struct
        """
        bill_key = self._get_bill_key(bill_id.native)
        bill_bytes, exists = op.Box.get(bill_key)
        assert exists, "Bill does not exist"
        return Bill.from_bytes(bill_bytes)

    @abimethod(readonly=True)
    def get_member(self, bill_id: ARC4UInt64, member: Account) -> Member:
        """
        Get member details for a bill
        
        Args:
            bill_id: The bill ID
            member: Member address
            
        Returns:
            Member struct
        """
        member_key = self._get_member_key(bill_id.native, member)
        member_bytes, exists = op.Box.get(member_key)
        assert exists, "Member not found"
        return Member.from_bytes(member_bytes)

    @abimethod(readonly=True)
    def get_user_bills(self, user: Account) -> DynamicArray[ARC4UInt64]:
        """
        Get all bills for a user
        
        Args:
            user: User address
            
        Returns:
            Array of bill IDs
        """
        user_key = self._get_user_key(user)
        user_bills_bytes, exists = op.Box.get(user_key)
        
        if not exists:
            return DynamicArray[ARC4UInt64]()
        
        return DynamicArray[ARC4UInt64].from_bytes(user_bills_bytes)

    @abimethod(readonly=True)
    def get_bill_counter(self) -> ARC4UInt64:
        """Get current bill counter"""
        return ARC4UInt64(self.bill_counter)

    @abimethod()
    def cancel_bill(self, bill_id: ARC4UInt64) -> Bool:
        """
        Cancel a bill (only creator can cancel, only if no payments made)
        
        Args:
            bill_id: The bill ID to cancel
            
        Returns:
            Success status
        """
        bill_id_native = bill_id.native
        
        # Load bill
        bill_key = self._get_bill_key(bill_id_native)
        bill_bytes, exists = op.Box.get(bill_key)
        assert exists, "Bill does not exist"
        bill = Bill.from_bytes(bill_bytes)
        
        # Only creator can cancel
        assert Txn.sender == bill.creator, "Only creator can cancel bill"
        
        # Can only cancel if no payments made
        assert bill.settled_count == 0, "Cannot cancel bill with payments"
        
        # Mark as settled to prevent future payments
        bill.is_settled = Bool(True)
        op.Box.put(bill_key, bill.bytes)
        
        return Bool(True)

    @abimethod()
    def update_member_share(
        self,
        bill_id: ARC4UInt64,
        member: Account,
        new_share: ARC4UInt64,
    ) -> Bool:
        """
        Update a member's share (only creator, only if member hasn't paid)
        
        Args:
            bill_id: The bill ID
            member: Member address to update
            new_share: New share amount
            
        Returns:
            Success status
        """
        bill_id_native = bill_id.native
        new_share_native = new_share.native
        
        # Load bill
        bill_key = self._get_bill_key(bill_id_native)
        bill_bytes, exists = op.Box.get(bill_key)
        assert exists, "Bill does not exist"
        bill = Bill.from_bytes(bill_bytes)
        
        # Only creator can update
        assert Txn.sender == bill.creator, "Only creator can update shares"
        assert not bill.is_settled.native, "Cannot update settled bill"
        
        # Load member
        member_key = self._get_member_key(bill_id_native, member)
        member_bytes, member_exists = op.Box.get(member_key)
        assert member_exists, "Member not found"
        member_data = Member.from_bytes(member_bytes)
        
        # Can only update if member hasn't paid
        assert not member_data.paid.native, "Cannot update paid member"
        assert new_share_native > 0, "Share must be greater than 0"
        
        # Update total amount
        old_share = member_data.share
        bill.total_amount = bill.total_amount - old_share + new_share_native
        
        # Update member share
        member_data.share = new_share_native
        op.Box.put(member_key, member_data.bytes)
        op.Box.put(bill_key, bill.bytes)
        
        return Bool(True)

    @abimethod()
    def remove_member(
        self,
        bill_id: ARC4UInt64,
        member: Account,
    ) -> Bool:
        """
        Remove a member from bill (only creator, only if member hasn't paid)
        
        Args:
            bill_id: The bill ID
            member: Member address to remove
            
        Returns:
            Success status
        """
        bill_id_native = bill_id.native
        
        # Load bill
        bill_key = self._get_bill_key(bill_id_native)
        bill_bytes, exists = op.Box.get(bill_key)
        assert exists, "Bill does not exist"
        bill = Bill.from_bytes(bill_bytes)
        
        # Only creator can remove
        assert Txn.sender == bill.creator, "Only creator can remove members"
        assert not bill.is_settled.native, "Cannot modify settled bill"
        assert bill.member_count > 1, "Cannot remove last member"
        
        # Load member
        member_key = self._get_member_key(bill_id_native, member)
        member_bytes, member_exists = op.Box.get(member_key)
        assert member_exists, "Member not found"
        member_data = Member.from_bytes(member_bytes)
        
        # Can only remove if member hasn't paid
        assert not member_data.paid.native, "Cannot remove paid member"
        
        # Update bill
        bill.total_amount -= member_data.share
        bill.member_count -= 1
        op.Box.put(bill_key, bill.bytes)
        
        # Delete member box
        deleted = op.Box.delete(member_key)
        assert deleted, "Failed to delete member box"
        
        # Remove bill from user's list
        self._remove_bill_from_user(member, bill_id_native)
        
        return Bool(True)

    @abimethod()
    def add_member(
        self,
        bill_id: ARC4UInt64,
        member: Account,
        share: ARC4UInt64,
        mbr_payment: gtxn.PaymentTransaction,
    ) -> Bool:
        """
        Add a new member to existing bill (only creator, only if bill not settled)
        
        Args:
            bill_id: The bill ID
            member: New member address
            share: Share amount for new member
            mbr_payment: Payment for box storage MBR
            
        Returns:
            Success status
        """
        bill_id_native = bill_id.native
        share_native = share.native
        
        # Verify MBR payment
        assert mbr_payment.receiver == Global.current_application_address, "MBR payment must go to app"
        
        # Load bill
        bill_key = self._get_bill_key(bill_id_native)
        bill_bytes, exists = op.Box.get(bill_key)
        assert exists, "Bill does not exist"
        bill = Bill.from_bytes(bill_bytes)
        
        # Only creator can add
        assert Txn.sender == bill.creator, "Only creator can add members"
        assert not bill.is_settled.native, "Cannot modify settled bill"
        assert share_native > 0, "Share must be greater than 0"
        
        # Check member doesn't already exist
        member_key = self._get_member_key(bill_id_native, member)
        existing_member_bytes, member_exists = op.Box.get(member_key)
        assert not member_exists, "Member already exists"
        
        # Create member box
        new_member = Member(
            share=share_native,
            paid=Bool(False),
        )
        op.Box.put(member_key, new_member.bytes)
        
        # Update bill
        bill.total_amount += share_native
        bill.member_count += 1
        op.Box.put(bill_key, bill.bytes)
        
        # Add bill to user's participation list
        self._add_bill_to_user(member, bill_id_native)
        
        return Bool(True)

    @subroutine
    def _get_bill_key(self, bill_id: UInt64) -> Bytes:
        """Generate bill box key"""
        return Bytes(b"bill_") + op.itob(bill_id)

    @subroutine
    def _get_member_key(self, bill_id: UInt64, member: Account) -> Bytes:
        """Generate member box key"""
        return Bytes(b"bill_") + op.itob(bill_id) + Bytes(b"_member_") + member.bytes

    @subroutine
    def _get_user_key(self, user: Account) -> Bytes:
        """Generate user participation box key"""
        return Bytes(b"user_") + user.bytes

    @subroutine
    def _add_bill_to_user(self, user: Account, bill_id: UInt64) -> None:
        """Add bill ID to user's participation list"""
        user_key = self._get_user_key(user)
        user_bills_bytes, exists = op.Box.get(user_key)
        
        if exists:
            # Append to existing list
            user_bills = DynamicArray[ARC4UInt64].from_bytes(user_bills_bytes)
            user_bills.append(ARC4UInt64(bill_id))
            op.Box.put(user_key, user_bills.bytes)
        else:
            # Create new list
            user_bills = DynamicArray[ARC4UInt64](ARC4UInt64(bill_id))
            op.Box.put(user_key, user_bills.bytes)

    @subroutine
    def _remove_bill_from_user(self, user: Account, bill_id: UInt64) -> None:
        """Remove bill ID from user's participation list"""
        user_key = self._get_user_key(user)
        user_bills_bytes, exists = op.Box.get(user_key)
        
        if exists:
            user_bills = DynamicArray[ARC4UInt64].from_bytes(user_bills_bytes)
            
            # Find and remove the bill_id
            for i in urange(user_bills.length):
                if user_bills[i].native == bill_id:
                    # Create new array without this bill
                    new_bills = DynamicArray[ARC4UInt64]()
                    for j in urange(user_bills.length):
                        if j != i:
                            new_bills.append(user_bills[j])
                    
                    if new_bills.length > 0:
                        op.Box.put(user_key, new_bills.bytes)
                    else:
                        # Delete box if empty
                        deleted = op.Box.delete(user_key)
                        assert deleted, "Failed to delete user box"
                    break

File: algosplit_frontend/lib/contract.ts
import algosdk from 'algosdk';
import { algodClient, APP_ID } from './algorand';

export interface Bill {
  creator: string;
  totalAmount: bigint;
  memberCount: bigint;
  settledCount: bigint;
  isSettled: boolean;
}

export interface Member {
  share: bigint;
  paid: boolean;
}

// Box MBR calculation (2500 + 400 per byte)
const BOX_FLAT_MIN_BALANCE = 2500;
const BOX_BYTE_MIN_BALANCE = 400;

export function calculateBoxMBR(boxSize: number): number {
  return BOX_FLAT_MIN_BALANCE + (boxSize * BOX_BYTE_MIN_BALANCE);
}

// Get current bill counter from contract
export async function getBillCounter(): Promise<number> {
  try {
    const appInfo = await algodClient.getApplicationByID(APP_ID).do();
    const globalState = appInfo.params.globalState;
    
    // Find bill_counter in global state
    const billCounterKey = Buffer.from('bill_counter').toString('base64');
    const billCounterState = globalState?.find((item) => {
      const keyStr = Buffer.from(item.key).toString('base64');
      return keyStr === billCounterKey;
    });
    
    if (billCounterState && billCounterState.value.type === 2) {
      return Number(billCounterState.value.uint);
    }
    
    return 0;
  } catch (error) {
    console.error('Error getting bill counter:', error);
    return 0;
  }
}

// Create bill transaction
export async function createBillTransaction(
  sender: string,
  members: string[],
  shares: number[], // in microAlgos
): Promise<algosdk.Transaction[]> {
  const suggestedParams = await algodClient.getTransactionParams().do();
  
  // Calculate proper MBR for boxes
  // Box MBR = 2500 + 400 * box_size (in microAlgos)
  const BOX_FLAT_MBR = 2500;
  const BOX_BYTE_MBR = 400;
  
  // Bill box: 32 (Account) + 8 (UInt64) + 8 (UInt64) + 8 (UInt64) + 1 (Bool) = 57 bytes
  const billBoxSize = 57;
  const billBoxMBR = BOX_FLAT_MBR + (BOX_BYTE_MBR * billBoxSize);
  
  // Member box: 8 (UInt64) + 1 (Bool) = 9 bytes
  const memberBoxSize = 9;
  const memberBoxMBR = BOX_FLAT_MBR + (BOX_BYTE_MBR * memberBoxSize);
  
  // User box: For each member, estimate minimal growth
  // New box: 2500 + 400 * (2 + 8) = 6500 microAlgos (1 bill)
  const userBoxMBR = BOX_FLAT_MBR + (BOX_BYTE_MBR * (2 + 8)); // Just 1 bill worth
  
  // Total MBR: 1 bill box + N member boxes + N user boxes
  // Add extra buffer for account minimum balance increase
  const boxesMBR = billBoxMBR + (members.length * (memberBoxMBR + userBoxMBR));
  const bufferMBR = 150000; // 0.15 ALGO buffer for account min balance
  const totalMBR = boxesMBR + bufferMBR;
  
  console.log('MBR Calculation:');
  console.log('Bill box MBR:', billBoxMBR / 1_000_000, 'ALGO');
  console.log('Member box MBR (each):', memberBoxMBR / 1_000_000, 'ALGO');
  console.log('User box MBR (each):', userBoxMBR / 1_000_000, 'ALGO');
  console.log('Boxes MBR:', boxesMBR / 1_000_000, 'ALGO');
  console.log('Buffer MBR:', bufferMBR / 1_000_000, 'ALGO');
  console.log('Total MBR:', totalMBR / 1_000_000, 'ALGO');

  // MBR payment transaction
  const mbrPayment = algosdk.makePaymentTxnWithSuggestedParamsFromObject({
    sender,
    receiver: algosdk.getApplicationAddress(APP_ID),
    amount: totalMBR,
    suggestedParams,
  });

  // Create ABI method for proper encoding
  const createBillMethod = new algosdk.ABIMethod({
    name: 'create_bill',
    args: [
      { type: 'address[]', name: 'members' },
      { type: 'uint64[]', name: 'shares' },
      { type: 'pay', name: 'mbr_payment' },
    ],
    returns: { type: 'uint64' },
  });

  // Encode method arguments
  const methodSelector = createBillMethod.getSelector();
  
  // Encode members array
  const membersType = algosdk.ABIType.from('address[]');
  const membersEncoded = membersType.encode(
    members.map(addr => algosdk.decodeAddress(addr).publicKey)
  );

  // Encode shares array
  const sharesType = algosdk.ABIType.from('uint64[]');
  const sharesEncoded = sharesType.encode(shares.map(s => BigInt(s)));

  // Get current bill counter to predict next bill ID
  const currentCounter = await getBillCounter();
  const nextBillId = currentCounter + 1;
  
  // Helper function to encode uint64 as big-endian bytes
  const encodeUint64 = (value: number): Uint8Array => {
    const bytes = new Uint8Array(8);
    new DataView(bytes.buffer).setBigUint64(0, BigInt(value), false);
    return bytes;
  };
  
  // Create box references
  const boxes: algosdk.BoxReference[] = [];
  
  // Bill box: "bill_<id>"
  const billBoxName = new Uint8Array([
    ...Buffer.from('bill_'),
    ...encodeUint64(nextBillId),
  ]);
  boxes.push({ appIndex: APP_ID, name: billBoxName });
  
  // Member boxes: "bill_<id>_member_<address>"
  for (const memberAddr of members) {
    const memberBoxName = new Uint8Array([
      ...Buffer.from('bill_'),
      ...encodeUint64(nextBillId),
      ...Buffer.from('_member_'),
      ...algosdk.decodeAddress(memberAddr).publicKey,
    ]);
    boxes.push({ appIndex: APP_ID, name: memberBoxName });
  }
  
  // User boxes: "user_<address>"
  for (const memberAddr of members) {
    const userBoxName = new Uint8Array([
      ...Buffer.from('user_'),
      ...algosdk.decodeAddress(memberAddr).publicKey,
    ]);
    boxes.push({ appIndex: APP_ID, name: userBoxName });
  }

  // Create application call
  const appCall = algosdk.makeApplicationNoOpTxnFromObject({
    sender,
    appIndex: APP_ID,
    appArgs: [
      methodSelector,
      membersEncoded,
      sharesEncoded,
    ],
    boxes,
    suggestedParams,
  });

  // Group transactions
  const txns = [mbrPayment, appCall];
  algosdk.assignGroupID(txns);

  return txns;
}

// Pay bill transaction
export async function payBillTransaction(
  sender: string,
  billId: number,
  billCreator: string,
  amount: number, // in microAlgos
): Promise<algosdk.Transaction[]> {
  const suggestedParams = await algodClient.getTransactionParams().do();

  // Payment to bill creator
  const payment = algosdk.makePaymentTxnWithSuggestedParamsFromObject({
    sender,
    receiver: billCreator,
    amount,
    suggestedParams,
  });

  // Create ABI method
  const payBillMethod = new algosdk.ABIMethod({
    name: 'pay_bill',
    args: [
      { type: 'uint64', name: 'bill_id' },
      { type: 'pay', name: 'payment' },
    ],
    returns: { type: 'bool' },
  });

  // Encode method arguments
  const methodSelector = payBillMethod.getSelector();
  const billIdType = algosdk.ABIType.from('uint64');
  const billIdEncoded = billIdType.encode(BigInt(billId));

  // Helper function to encode uint64 as big-endian bytes
  const encodeUint64 = (value: number): Uint8Array => {
    const bytes = new Uint8Array(8);
    new DataView(bytes.buffer).setBigUint64(0, BigInt(value), false);
    return bytes;
  };

  // Create box references for the boxes that will be accessed
  const boxes: algosdk.BoxReference[] = [];
  
  // Bill box: "bill_<id>"
  const billBoxName = new Uint8Array([
    ...Buffer.from('bill_'),
    ...encodeUint64(billId),
  ]);
  boxes.push({ appIndex: APP_ID, name: billBoxName });
  
  // Member box: "bill_<id>_member_<address>"
  const memberBoxName = new Uint8Array([
    ...Buffer.from('bill_'),
    ...encodeUint64(billId),
    ...Buffer.from('_member_'),
    ...algosdk.decodeAddress(sender).publicKey,
  ]);
  boxes.push({ appIndex: APP_ID, name: memberBoxName });

  // Application call
  const appCall = algosdk.makeApplicationNoOpTxnFromObject({
    sender,
    appIndex: APP_ID,
    appArgs: [
      methodSelector,
      billIdEncoded,
    ],
    boxes,
    suggestedParams,
  });

  // Group transactions
  const txns = [payment, appCall];
  algosdk.assignGroupID(txns);

  return txns;
}

// Get bill details by reading box from indexer
export async function getBill(billId: number): Promise<Bill | null> {
  try {
    const billBoxName = new Uint8Array([
      ...Buffer.from('bill_'),
      ...algosdk.encodeUint64(billId),
    ]);
    
    const boxResponse = await algodClient.getApplicationBoxByName(APP_ID, billBoxName).do();
    const boxValue = boxResponse.value;
    
    // Parse Bill struct: (address, uint64, uint64, uint64, bool)
    // address: 32 bytes, uint64: 8 bytes each, bool: 1 byte
    const creator = algosdk.encodeAddress(new Uint8Array(boxValue.slice(0, 32)));
    const totalAmount = BigInt('0x' + Buffer.from(boxValue.slice(32, 40)).toString('hex'));
    const memberCount = BigInt('0x' + Buffer.from(boxValue.slice(40, 48)).toString('hex'));
    const settledCount = BigInt('0x' + Buffer.from(boxValue.slice(48, 56)).toString('hex'));
    const isSettled = boxValue[56] !== 0;

    return {
      creator,
      totalAmount,
      memberCount,
      settledCount,
      isSettled,
    };
  } catch (error) {
    console.error('Error getting bill:', error);
    return null;
  }
}

// Get member details by reading box from indexer
export async function getMember(billId: number, memberAddress: string): Promise<Member | null> {
  try {
    const memberBoxName = new Uint8Array([
      ...Buffer.from('bill_'),
      ...algosdk.encodeUint64(billId),
      ...Buffer.from('_member_'),
      ...algosdk.decodeAddress(memberAddress).publicKey,
    ]);
    
    const boxResponse = await algodClient.getApplicationBoxByName(APP_ID, memberBoxName).do();
    const boxValue = boxResponse.value;
    
    // Parse Member struct: (uint64, bool)
    // uint64: 8 bytes, bool: 1 byte
    const share = BigInt('0x' + Buffer.from(boxValue.slice(0, 8)).toString('hex'));
    const paid = boxValue[8] !== 0;

    return { share, paid };
  } catch (error: any) {
    // Box not found is expected if the user is the creator (not a member)
    if (error?.message?.includes('box not found') || error?.status === 404) {
      // Silently return null - this is expected for bill creators
      return null;
    }
    console.error('Error getting member:', error);
    return null;
  }
}

// Get user bills by reading box from indexer
export async function getUserBills(userAddress: string): Promise<number[]> {
  try {
    const userBoxName = new Uint8Array([
      ...Buffer.from('user_'),
      ...algosdk.decodeAddress(userAddress).publicKey,
    ]);
    
    // First check if the box exists to avoid 404 errors in console
    try {
      const boxes = await algodClient.getApplicationBoxes(APP_ID).do();
      const boxNameBase64 = Buffer.from(userBoxName).toString('base64');
      const boxExists = boxes.boxes.some((box: any) => 
        Buffer.from(box.name).toString('base64') === boxNameBase64
      );
      
      if (!boxExists) {
        // Box doesn't exist - user hasn't been added to any bills as a member
        return [];
      }
    } catch (checkError) {
      // If we can't check boxes, try to fetch anyway
      console.error('Error checking boxes:', checkError);
    }
    
    const boxResponse = await algodClient.getApplicationBoxByName(APP_ID, userBoxName).do();
    const boxValue = boxResponse.value;
    
    // Parse dynamic array of uint64
    // First 2 bytes: array length
    const length = (boxValue[0] << 8) | boxValue[1];
    const billIds: number[] = [];
    
    for (let i = 0; i < length; i++) {
      const offset = 2 + (i * 8);
      const billId = Number(BigInt('0x' + Buffer.from(boxValue.slice(offset, offset + 8)).toString('hex')));
      billIds.push(billId);
    }

    return billIds;
  } catch (error: any) {
    // Box not found is expected if user hasn't been added to any bills as a member
    // This happens when user only created bills but isn't a member of any
    if (error?.message?.includes('box not found') || error?.status === 404) {
      // Silently return empty array - this is expected behavior
      return [];
    }
    // Only log unexpected errors
    console.error('Error getting user bills:', error);
    return [];
  }
}

// Get all bills where user is the creator
export async function getBillsCreatedBy(creatorAddress: string): Promise<number[]> {
  try {
    // Get current bill counter
    const currentCounter = await getBillCounter();
    const billIds: number[] = [];
    
    // Check each bill to see if this user is the creator
    for (let i = 1; i <= currentCounter; i++) {
      const bill = await getBill(i);
      if (bill && bill.creator === creatorAddress) {
        billIds.push(i);
      }
    }
    
    return billIds;
  } catch (error) {
    console.error('Error getting bills created by user:', error);
    return [];
  }
}

// Get all member addresses for a bill by listing boxes
export async function getBillMembers(billId: number): Promise<string[]> {
  try {
    // Get all boxes for the application
    const boxes = await algodClient.getApplicationBoxes(APP_ID).do();
    
    // Filter boxes that match pattern: bill_<id>_member_<address>
    const billPrefix = Buffer.concat([
      Buffer.from('bill_'),
      Buffer.from(algosdk.encodeUint64(billId)),
      Buffer.from('_member_'),
    ]);
    
    const memberAddresses: string[] = [];
    
    for (const box of boxes.boxes) {
      const boxName = Buffer.from(box.name);
      
      // Check if this box is a member box for this bill
      if (boxName.length === billPrefix.length + 32 && 
          boxName.subarray(0, billPrefix.length).equals(billPrefix)) {
        // Extract the 32-byte address from the box name
        const addressBytes = boxName.subarray(billPrefix.length);
        const address = algosdk.encodeAddress(addressBytes);
        memberAddresses.push(address);
      }
    }
    
    return memberAddresses;
  } catch (error) {
    console.error('Error getting bill members:', error);
    return [];
  }
}

// Cancel bill transaction
export async function cancelBillTransaction(
  sender: string,
  billId: number,
): Promise<algosdk.Transaction> {
  const suggestedParams = await algodClient.getTransactionParams().do();

  // Create ABI method
  const cancelBillMethod = new algosdk.ABIMethod({
    name: 'cancel_bill',
    args: [{ type: 'uint64', name: 'bill_id' }],
    returns: { type: 'bool' },
  });

  const methodSelector = cancelBillMethod.getSelector();
  const billIdType = algosdk.ABIType.from('uint64');
  const billIdEncoded = billIdType.encode(BigInt(billId));

  // Generate bill box key
  const billIdBytes = new Uint8Array(8);
  new DataView(billIdBytes.buffer).setBigUint64(0, BigInt(billId), false);
  const billBoxName = new Uint8Array([
    ...Buffer.from('bill_'),
    ...billIdBytes,
  ]);

  return algosdk.makeApplicationNoOpTxnFromObject({
    sender,
    appIndex: APP_ID,
    appArgs: [
      methodSelector,
      billIdEncoded,
    ],
    boxes: [
      {
        appIndex: APP_ID,
        name: billBoxName,
      },
    ],
    suggestedParams,
  });
}
